### PR TITLE
Add `source type` and support for downloading `source content` in Application

### PR DESCRIPTION
**Pull request by Revolyssup**
**Description**

This PR fixes #

**Notes for Reviewers**

**[Signed commits](https://github.com/meshery/meshery/blob/master/CONTRIBUTING.md#signing-off-on-commits-developer-certificate-of-origin)**
- [ ] Yes, I signed my commits.

### CODE REVIEW

File: handlers/meshery_application_handler.go [hunk 1]
```diff
@@ -122,17 +125,43 @@ func (h *Handler) handleApplicationPOST(
 		mesheryApplication := parsedBody.ApplicationData
 
 		bytApplication := []byte(mesheryApplication.ApplicationFile)
-		// check whether the uploaded file is a docker compose file
-		err := kompose.IsManifestADockerCompose(bytApplication, "")
-		if err == nil {
-			res, err := kompose.Convert(bytApplication) // convert the docker compose file into kubernetes manifest
+		mesheryApplication.SourceContent = bytApplication
+		var composeerr = fmt.Errorf("not docker compose")
+		if sourcetype == "" {
```
Comment (sudo-NithishKarthik) [inline handlers/meshery_application_handler.go:130]: Do we need to support this?
This still makes the file type inference ambiguous. Should we not make it a required thing?

Comment (Revolyssup) [reply handlers/meshery_application_handler.go:130]: The spec states that it should be optional. Let's carry this discussion on the spec.

Comment (leecalcote) [reply handlers/meshery_application_handler.go:130]: Spec update: Given our outstanding challenges, this parameter is to be required.

Comment (Revolyssup) [reply handlers/meshery_application_handler.go]: @leecalcote  thank you for clarification. I have made the query param required now.

File: handlers/meshery_application_handler.go [hunk 2]
```diff
@@ -88,10 +91,13 @@ func (h *Handler) handleApplicationPOST(
 	defer func() {
 		_ = r.Body.Close()
 	}()
-
+	sourcetype := r.URL.Query().Get("source-type")
```
[the remark this answers is not in the document]
Comment (leecalcote) [reply handlers/meshery_application_handler.go:94]: `/download`

File: models/meshery_application.go [hunk 3]
```diff
@@ -15,9 +27,10 @@ type MesheryApplication struct {
...
 	// but the remote provider is allowed to provide one
-	UserID   *string `json:"user_id" gorm:"-"`
-	Location sql.Map `json:"location"`
-
-	UpdatedAt *time.Time `json:"updated_at,omitempty"`
-	CreatedAt *time.Time `json:"created_at,omitempty"`
+	UserID        *string         `json:"user_id" gorm:"-"`
+	Location      sql.Map         `json:"location"`
+	Type          ApplicationType `json:"type"`
+	SourceContent []byte
```
Comment (MUzairS15) [inline models/meshery_application.go:33]: @Revolyssup 
SourceContent []byte `json: "source_content"` ??

**Review by Chadha93 — COMMENTED** [review]
@Revolyssup Could you please merge this cloud PR by @MUzairS15, it would help me to run and test my UI changes directly https://github.com/layer5io/meshery-cloud/pull/233 and resolve any conflicts if any. I'll be pushing more changes for UI. Source-type import completed.

**Comment by Chadha93** [comment]
@Revolyssup hardcoding removed.
apps

**Comment by leecalcote** [comment]
Merge Conflict

**Comment by leecalcote** [comment]
Feature completed? Demo on today's call?

**Comment by Revolyssup** [comment]
> Feature completed? Demo on today's call?

There were three enhancements noted by @MUzairS15 which are pending and would be done later by tomorrow. But if @Chadha93  has changed the UI hardcoding then we can demo it.

**Comment by leecalcote** [comment]
Lint check love needed.

File: ui/components/UploadImport.js [hunk 4]
```diff
@@ -51,21 +51,29 @@ const styles = makeStyles(() => ({
...
+const UploadImport = ({ handleUpload, handleImport, configuration, isApplication, supportedTypes }) => {
   const classes = styles();
   const [open, setOpen] = React.useState(false);
   const [input, setInput] = React.useState();
   const [isError, setIsError] = React.useState(false);
+  const [fileType, setFileType] = React.useState();
+  const [sourceType, setSourceType] = React.useState();
 
-
+  const handleFilteType = (index) => {
```
Comment (warunicorn19) [inline ui/components/UploadImport.js:62]: best to use a proper name for the function

File: ui/components/UploadImport.js [hunk 5]
```diff
@@ -51,21 +51,29 @@ const styles = makeStyles(() => ({
...
+    setFileType(supportedTypes[index]?.supported_extensions);
+    setSourceType(supportedTypes[index]?.application_type);
+  }
+  console.log("POI",supportedTypes);
```
Comment (warunicorn19) [inline ui/components/UploadImport.js:66]:
```suggestion
```

File: ui/components/UploadImport.js [hunk 6]
```diff
@@ -75,15 +83,17 @@ const UploadImport = ({ handleUpload, handleImport, configuration }) => {
   };
 
   const handleSubmit = () => {
-    handleUpload(input)
+    console.log("QAZWSX", sourceType);
```
Comment (warunicorn19) [inline ui/components/UploadImport.js:86]:
```suggestion
```

**Review by warunicorn19 — CHANGES_REQUESTED** [review]

File: ui/components/UploadImport.js (hunk 5 above)
Comment (Chadha93) [reply ui/components/UploadImport.js]: Fixed

File: ui/components/UploadImport.js (hunk 6 above)
Comment (Chadha93) [reply ui/components/UploadImport.js]: Fixed

File: ui/components/UploadImport.js (hunk 4 above)
Comment (Chadha93) [reply ui/components/UploadImport.js]: Thank you, fixed.

**Comment by Chadha93** [comment]
> Lint check love needed.

Done for the UI part now.

**Comment by leecalcote** [comment]
1. API docs in Meshery Docs
2. API Endpoint spreadsheet update
3. Ensure that the GraphQL subscription to view Applications includes `source-type`. 
4. Ensure that the GraphQL GET to view Applications includes `source-type`. 
----
5. Confusion over REST API and GraphQL API.
6. Query string vs POST body - inconsistency.

**Comment by leecalcote** [comment]
Test failure:

```
--- FAIL: TestAppView (0.05s)
[276](https://github.com/meshery/meshery/runs/7503460408?check_suite_focus=true#step:5:277)
    --- FAIL: TestAppView/View_Applications_with_Name (0.01s)
```

**Comment by leecalcote** [comment]
Screen Shot 2022-07-28 at 2 19 30 PM

**Comment by leecalcote** [comment]
Screen Shot 2022-07-28 at 2 20 43 PM

**Comment by Chadha93** [comment]
> Screen Shot 2022-07-28 at 2 20 43 PM

It is intentional, we are not allowing users to upload(Browse) helm charts. Only URL support. cc @Revolyssup

**Comment by Revolyssup** [comment]
> If Helm Chart isn't currently supported, then disable it as a selection item.
> 
> Screen Shot 2022-07-28 at 2 20 43 PM

Helm is supported but only through URL, that is why the browse has been disabled.

**Comment by leecalcote** [comment]
Undeploy of an application produces this server log:

Screen Shot 2022-07-28 at 2 24 02 PM

```INFO[0482] pattern resource successfully deleted from remote provider```

This shouldn't be a `delete`, but an `undeploy`.

**Comment by leecalcote** [comment]
> > If Helm Chart isn't currently supported, then disable it as a selection item.
> > Screen Shot 2022-07-28 at 2 20 43 PM
> 
> Helm is supported but only through URL, that is why the browse has been disabled.

Then, disable the ```File``` input field, too.

**Comment by leecalcote** [comment]
The red color used for Undeploy in the grid view of Applications differs from the red color used for the Undeploy button in the deployment modal. Please make consistent by using the color from the grid view.

**Comment by leecalcote** [comment]
Please make "Deploy" the primary action color in the Deployment modal (not light gray).
Screen Shot 2022-07-28 at 2 27 18 PM

**Comment by Chadha93** [comment]
> The red color used for Undeploy in the grid view of Applications differs from the red color used for the Undeploy button in the deployment modal. Please make consistent by using the color from the grid view.

https://github.com/meshery/meshery/pull/5842#pullrequestreview-1054639526

**Comment by leecalcote** [comment]
Helm logo missing.

Screen Shot 2022-07-28 at 2 32 03 PM

**Comment by Chadha93** [comment]
> Helm logo missing.
> 
> Screen Shot 2022-07-28 at 2 32 03 PM

Please upload a new Helm file older one won't have source-type/content. 
https://github.com/prometheus-community/helm-charts/releases/download/alertmanager-0.18.3/alertmanager-0.18.3.tgz

**Comment by leecalcote** [comment]
1. Please include thee Application name in this notification. 
1. Lowercase all words other than the title of the Application, which should be put in double quotation marks.
1. Stop yelling at the user with an exclamation mark and remove end of sentence punctuation instead! (yes, that exclamation mark was a dose of our own medicine)

Screen Shot 2022-07-28 at 2 37 28 PM

**Comment by leecalcote** [comment]
> > Helm logo missing.
> > Screen Shot 2022-07-28 at 2 32 03 PM
> 
> Please upload a new Helm file older one won't have source-type/content. https://github.com/prometheus-community/helm-charts/releases/download/alertmanager-0.18.3/alertmanager-0.18.3.tgz

Have a broken image link isn't the answer to not supporting a file type.

I misspoke: the Applications in this screenshot are either a docker compose or a kubernetes manifest, not Helm Chart.

It seems there is more than one bug here.

**Comment by MUzairS15** [comment]
Screenshot 2022-07-29 at 1 53 03 AM


> > > Helm logo missing.
> > > Screen Shot 2022-07-28 at 2 32 03 PM
> > 
> > 
> > Please upload a new Helm file older one won't have source-type/content. [prometheus-community/helm-charts@`alertmanager-0.18.3` alertmanager-0.18.3.tgz (download)](https://github.com/prometheus-community/helm-charts/releases/download/alertmanager-0.18.3/alertmanager-0.18.3.tgz)
> 
> Have a broken image link isn't the answer to not supporting a file type.
> 
> I misspoke: the Applications in this screenshot are either a docker compose or a kubernetes manifest, not Helm Chart.
> 
> It seems there is more than one bug here.

Working for me, can you rebuild and try again

**Comment by Chadha93** [comment]
Field label - 
image

**Comment by leecalcote** [comment]
@MUzairS15 funky. Yes, I can. Although, before I do, will someone cleanup the Docker Compose logo? There's extra padding that is preventing it from being the same size as the other logos.

**Comment by Chadha93** [comment]
> @MUzairS15 funky. Yes, I can. Although, before I do, will someone cleanup the Docker Compose logo? There's extra padding that is preventing it from being the same size as the other logos.

image

**Comment by Chadha93** [comment]
- [x] Application name on deploy/undeploy action
- [x] lowercase all letters 
- [x] no yelling at the user, we are polite. 
Screenshot from 2022-07-29 03-22-55
  // @leecalcote

**Comment by leecalcote** [comment]
> * [x]  Application name on deploy/undeploy action
> * [x]  lowercase all letters
> * [x]  no yelling at the user, we are polite.
>   Screenshot from 2022-07-29 03-22-55
>   // @leecalcote

Ha! 😂 I suppose I more than deserve a rash of sarcasm. 😆

File: ui/utils/fileDownloader.js [hunk 7]
```diff
@@ -0,0 +1,12 @@
+export function fileDownloader(id, name, source_type) {
+  let dataStr = JSON.stringify(id);
+  let dataUri = `/api/application/download/${id}?source-type=${source_type}`+ encodeURIComponent(dataStr);
```
Comment (leecalcote) [inline ui/utils/fileDownloader.js:3]: Are we still using query params or am I looking at an old commit?

File: ui/lib/data-fetch.js [hunk 8]
```diff
@@ -28,11 +28,11 @@ const dataFetch = (url, options = {}, successFn, errorFn) => {
 
         return result;
       } else {
-        throw res
+        throw res.text()
       }
 
     }).then(successFn)
-    .catch((res) => res.text().then(errorFn));
+    .catch((e) => e.then(text => errorFn(text)));
```
Comment (Chadha93) [inline ui/lib/data-fetch.js:35]: @Abhishek-kumar09 Here one integration test is failing `errorFn` Any suggestion?

**Comment by Chadha93** [comment]
All the above suggested UI changes are done here, @MUzairS15 you can test the PR now, could you please update the query parameter issue? PR should be good to merge then.

File: mesheryctl/internal/cli/root/app/onboard.go [hunk 9]
```diff
@@ -22,14 +22,17 @@ import (
...
 
 var (
-	skipSave bool // skip saving a app
-	appFile  string
+	skipSave   bool   // skip saving the app
+	appFile    string // app file
+	sourceType string // app file type (manifest / compose)
 )
 
+var validSourceTypes = []string{"compose", "manifest", "helm"}
```
Comment (MUzairS15) [inline mesheryctl/internal/cli/root/app/onboard.go:30]: @gr455 These are not the source-types, we are using, Refer the doc. // @Revolyssup 
Also, these should not be hardocded. 
You can use `/api/application/types` endpoint to get valid sourcetypes. 
In `PreRunE` step you can get all the types and use.

**Comment by Revolyssup** [comment]
@leecalcote  Lets merge this. `mesheryctl` consideration is left. But that can be carried out in a separate PR. The query parameter has been removed.

**Comment by Revolyssup** [comment]
@MUzairS15  Can you sync this with upstream? There are few conflicts in UI.

**Comment by gr455** [comment]
> @leecalcote Lets merge this. `mesheryctl` consideration is left. But that can be carried out in a separate PR. The query parameter has been removed.

This is done now

**Comment by leecalcote** [comment]
Merge conflicts....

File: ui/lib/data-fetch.js (hunk 8 above)
Comment (Chadha93) [reply ui/lib/data-fetch.js:35]: @Abhishek-kumar09 any quick suggestions to clear this check?

**Comment by leecalcote** [comment]
139 commits in this PR. I suspect this branch needs refreshed with changes in master...

**Comment by Chadha93** [comment]
Yay!! Finally Merged.

**Pull request merged**